### DISCUSSION TITLE
Navigate does not clear search by default

### DISCUSSION THREAD

**Discussion by joakimgunst** [TanStack/router#242] [General]
I'm not sure this is a bug or an intentional change. I did not find any mention about it in the release notes. When upgrading from 3.3.4 to 3.7.0, I noticed that search params are no longer cleared when calling `navigate` without a `search` property.

For example:

```ts
const navigate = useNavigate();
const handleClick = (to: string) => navigate({ to });
```

will preserve the search params while

```ts
const handleClick = (to: string) => navigate({ to, search: {} });
```

will clear them.

The same thing happens with `Link` without a `search` prop.

This caused some unexpected behavior in our app.

My assumption would be that search params would be cleared unless explicitly preserved with:

```ts
const handleClick = (to: string) => navigate({ to, search: (search) => ({ ...search }) });
```

**Comment by tannerlinsley** [comment]
Can you make a codesandbox we can debug together?

**Comment by joakimgunst** [comment]
Here's an example: https://stackblitz.com/edit/react-ts-qjykxk?file=List.tsx

From a consistency perspective it makes sense that when `search` is undefined the current search is not changed. This is similar to how an undefined `to` does not change the current path. But in practice keeping the search parameters is almost never something I would want went navigating to a new page, so the behavior was surprising.

I suppose we could create wrappers for `useNavigate` and `Link` which set `search: {}` by default, to work around the issue.

Maybe it would be enough if the behavior was documented a bit more clearly.

And by the way, we're big fans of React Query and now also React Location. Thank you so much for these libraries!

**Reply by tannerlinsley** [reply]
I'm pushing a fix for this right now. Update your library in ~5 minutes :)

**Reply by tannerlinsley** [reply]
Github actions is having an outage. It will be a bit before it gets released. Sorry!

**Reply by tannerlinsley** [reply]
When it does release, this sandbox should work: https://stackblitz.com/edit/react-ts-3k668n?file=package.json

**Reply by joakimgunst** [reply]
Fix works great. Thank you for the quick response!

**Reply by abed-daloopa** [reply]
Hey @tannerlinsley I just tried the code in the sandbox and it's not getting cleared

https://github.com/user-attachments/assets/566eeb0e-a874-467b-a3e9-a1e4a177c508
